vendors/ti/boards/cc3220_launchpad/ports/common_io/iot_i2c.c

```diff
@@ -0,0 +1,391 @@
+            break;
+
+        case eI2CGetBusState:
+            *( IotI2CBusStatus_t * ) pvBuffer = _getState( pxI2CPeripheral );
+            break;
+
+        default:
+            ret = IOT_I2C_INVALID_VALUE;
+            break;
+    }
+
+    return ret;
+}
+
+int32_t iot_i2c_close( IotI2CHandle_t const pxI2CPeripheral )
+{
+    if( ( pxI2CPeripheral == NULL ) || ( pxI2CPeripheral->handle == NULL ) )
+    {
+        return IOT_I2C_INVALID_VALUE;
+    }
+
+    I2C_close( pxI2CPeripheral->handle );
+    /*pxI2CPeripheral->handle = NULL; */
+    memset( pxI2CPeripheral, 0, sizeof( IotI2CDescriptor_t ) );
+
+    return IOT_I2C_SUCCESS;
+}
+
+int32_t iot_i2c_cancel( IotI2CHandle_t const pxI2CPeripheral )
+{
+    if( ( pxI2CPeripheral == NULL ) || ( pxI2CPeripheral->handle == NULL ) )
+    {
+        return IOT_I2C_INVALID_VALUE;
+    }
+
+    I2C_cancel( pxI2CPeripheral->handle );
+
+    return IOT_I2C_SUCCESS;
+}
+
+static int32_t _doTransfer( IotI2CHandle_t const pxI2CPeripheral,
+                            uint8_t * const pvBuffer,
+                            size_t xBytes,
+                            uint8_t op,
+                            uint8_t mode )
+{
+    if( ( pxI2CPeripheral == NULL ) || ( pvBuffer == NULL ) || ( xBytes <= 0 ) )
+    {
+        return IOT_I2C_INVALID_VALUE;
+    }
+
+    if( pxI2CPeripheral->transaction.slaveAddress == 0 )
+    {
+        return IOT_I2C_SLAVE_ADDRESS_NOT_SET;
+    }
+
+    if( _getState( pxI2CPeripheral ) == eI2cBusBusy )
+    {
+        return IOT_I2C_BUSY;
+    }
+
+    int32_t ret = IOT_I2C_SUCCESS;
+    bool status = false;
+
+    if( op == _OP_READ )
+    {
+        pxI2CPeripheral->transaction.readBuf = pvBuffer;
+        pxI2CPeripheral->transaction.readCount = xBytes;
+    }
+    else
+    {
+        pxI2CPeripheral->transaction.writeBuf = pvBuffer;
+        pxI2CPeripheral->transaction.writeCount = xBytes;
+    }
+
+    /* If sending STOP condition. */
+    if( _isFlagNotSet( pxI2CPeripheral, _FLAG_NO_STOP ) )
+    {
+        status = I2C_transfer( pxI2CPeripheral->handle,
+                               &pxI2CPeripheral->transaction );
+
+        if( status )
+        {
+            /* If sync, wait for the transfer to finish. */
+            if( ( mode == _MODE_SYNC ) && ( xSemaphoreTake( pxI2CPeripheral->semphr, pdMS_TO_TICKS( pxI2CPeripheral->config.ulMasterTimeout ) ) == pdFALSE ) )
+            {
+                /* If callback is not invoked with timeout, return timeout. */
+                ret = IOT_I2C_BUS_TIMEOUT;
+            }
+        }
+        else
+        {
+            ret = IOT_I2C_NACK;
+        }
+
+        pxI2CPeripheral->transaction.writeBuf = NULL;
+        pxI2CPeripheral->transaction.readBuf = NULL;
+    }
+    else
+    {
+        pxI2CPeripheral->flags &= ~_FLAG_NO_STOP;
+
+        /* If async, invoke user's callback. */
+        if( ( mode == _MODE_ASYNC ) && ( pxI2CPeripheral->userCallback != NULL ) )
```

Comment:
Is the expectation that the user callback is invoked when the i2c transaction is completed ?

Reply:
Question – are callbacks only for asynchronous operation?